Added Call class and re-implemented spy.getCall

jonathan-benn opened the pull request:
Hi Kir,

I added the Call class and did a little refactoring inside of Spy. Please let me know if you're unhappy with the refactoring and I can undo it or put it in a separate commit. There are some extra methods that need to be added to the Call class to match Sinon.JS, but I think that can wait until a future commit.

--Jonathan

Fixes note35/sinon#2
Fixes note35/sinon#3
Fixes note35/sinon#4

jonathan-benn commented:
I wonder if I picked the wrong name for the new class `Call`? Perhaps it should be `SpyCall` to better match Sinon.JS?

note35 commented on sinon/test/TestSinonSpy.py:
```diff
@@ -857,6 +857,74 @@ def test240_neverCalledWithMatch_args(self):
+        exception = BaseException()
+        try:
+            sinon.g.D_func(exception)
+        except BaseException, e:
```
Here is a syntax error in Python3. If I am correct, it should be:

```
except BaseException as e:
```

note35 commented on sinon/lib/call.py:
```diff
@@ -0,0 +1,23 @@
+'''
+Copyright (c) 2017, Kir Chou
```
You can write your name here! :)

jonathan-benn replied:
It's actually better for your project if the whole thing's copyright is under one person's name. In the future, it will be easier for you if you want to re-release the software under a different license, create a closed-source fork or whatever. Since Sinon.PY is being released under an open source license, I'm fine with assigning you the copyright.

I will add myself as an author though. Thanks!

note35 commented on sinon/lib/spy.py:
```diff
@@ -68,77 +70,61 @@ def __set_args_list(self, new_args_list):
-            return getattr(self.g, self.obj.__name__).error_list
-        elif self.args_type == "PURE": #TODO: consider to remove this condition
-            return getattr(self.pure, "func").error_list
+        #TODO: consider to remove "PURE" condition
```
move this comment to `def __get_wrapper(self):`

note35 commented on sinon/lib/spy.py:
```diff
@@ -68,77 +70,61 @@ def __set_args_list(self, new_args_list):
-            return getattr(self.g, self.obj.__name__).ret_list
-        elif self.args_type == "PURE": #TODO: consider to remove this condition
-            return getattr(self.pure, "func").ret_list
+        #TODO: consider to remove "PURE" condition
```
no need this comment anymore

note35 commented on sinon/lib/spy.py:
```diff
@@ -6,6 +6,8 @@
 from .util import CollectionHandler as uch
 from .base import SinonBase
 from .matcher import SinonMatcher, Matcher
+from .call import Call
```
I don't find any `Call()` declaration here, do you need to use `Call()` in spy?

jonathan-benn replied:
I originally put `Call()` in `spy`, but then moved the call to `wrapper`. I left this import line by accident. Good catch!

note35 commented on sinon/test/TestSinonSpy.py:
```diff
@@ -868,3 +936,4 @@ def test270_args_and_kwargs(self):
         sinon.g.E_func(1, 2, a=1, b=2)
         self.assertListEqual(spy.args, [(), (1,), (1,2)])
         self.assertListEqual(spy.kwargs, [{}, {'a':1}, {'a':1,'b':2}])
+
```
remove this line

note35 commented:
You did a great job! I am appreciated to your this big refactoring work. I will review it more carefully while this change passed(or failed in known test cases) the CI test.

1. Move getCall from base to spy.
Not a bad idea, I agree with this change.

2. Make new call class in lib.
Here are 2 suggestions:
    a. move `sinon/lib/call.py` to `sinon/lib/util/call.py` and modifying related `import`. 
    b. move `class Call` to `sinon/lib/util/Wrapper.py` without creating a new file.

3. Refactoring work.
Good Job!

4. Create call_list
It's a good idea, we need to consider remove `CALLQUEUE` in the following change.


>I wonder if I picked the wrong name for the new class Call? Perhaps it should be SpyCall to better match Sinon.JS?

It's a naming question, I personally prefer `SpyCall` for matching Sinon.JS's design.

note35 commented on sinon/lib/util/Wrapper.py:
```diff
@@ -2,6 +2,11 @@
 """
+import traceback
+import sys
+sys.path.insert(0, '../../')
```
This is a bad idea. Please read my comment [here](https://github.com/note35/sinon/pull/8#issuecomment-334007839)

note35 commented on sinon/lib/util/Wrapper.py:
```diff
@@ -2,6 +2,11 @@
+import traceback
+import sys
+sys.path.insert(0, '../../')
+from lib.call import Call
```
This will cause CI failed, fixing the folder structure might resolve this at the same time.

note35 commented on sinon/lib/call.py:
```diff
@@ -0,0 +1,23 @@
+        self.proxy = None
+        self.returnValue = None
+        self.stack = None
+
```
remove empty lines

jonathan-benn commented:
Hi Kir,

It looks like I fixed the Continuous Integration issue. I also implemented all of your suggestions, and chose option (2a): move `sinon/lib/call.py` to `sinon/lib/util/SpyCall.py`

Sincerely,

--Jonathan

note35 commented on sinon/lib/spy.py:
```diff
@@ -68,77 +69,60 @@ def __set_args_list(self, new_args_list):
-            getattr(self.pure, "func").__set__("args_list", new_args_list)
+            return getattr(self.pure, "func")
+        else:
+            raise Exception('Wrapper object cannot be found')
```
https://github.com/note35/sinon/blob/master/sinon/lib/util/ErrorHandler.py

I just lost this part, I recommend to use `ErrorHandler.py` to define exception. You can define error function such as below and import it to here.

```python
def wrapper_object_is_not_found_error():
    error_msg = 'Wrapper object cannot be found'
    return __exception_helper(error_msg)
```

Then call that function here:

```python
from .util import ErrorHandler
...
else:
    ErrorHandler.wrapper_object_is_not_found_error()
...
```
